services/httpd/handler.go

```diff
@@ -993,8 +1017,437 @@ func fixLiterals(node influxql.Node) {
 	}
 }
 
+type RetentionRule struct {
+	Type                      string `json:"type"`
+	EverySeconds              int64  `json:"everySeconds"`
+	ShardGroupDurationSeconds int64  `json:"shardGroupDurationSeconds"`
+}
+
+type BucketUpdate struct {
+	Description    string          `json:"description"`
+	Name           string          `json:"name"`
+	RetentionRules []RetentionRule `json:"retentionRules"`
+}
+
+// BucketsBody and RetentionRule should match the 2.0 API definition.
+type BucketsBody struct {
+	BucketUpdate
+	OrgID      string `json:"orgId"`
+	Rp         string `json:"rp"`
+	SchemaType string `json:"schemaType"`
+}
+
+type Bucket struct {
+	BucketsBody
+	ID                  string    `json:"id,omitempty"`
+	Type                string    `json:"type"`
+	RetentionPolicyName string    `json:"rp,omitempty"` // This to support v1 sources
+	CreatedAt           time.Time `json:"createdAt"`
+	UpdatedAt           time.Time `json:"updatedAt"`
+}
+
+func (h *Handler) ServePostBucketsV2(w http.ResponseWriter, r *http.Request, user meta.User) {
+	var bs []byte
+	if r.ContentLength > 0 {
+		if h.Config.MaxBodySize > 0 && r.ContentLength > int64(h.Config.MaxBodySize) {
+			h.httpError(w, http.StatusText(http.StatusRequestEntityTooLarge), http.StatusRequestEntityTooLarge)
+			return
+		}
+
+		// This will just be an initial hint for the reader, as the
+		// bytes.Buffer will grow as needed when ReadFrom is called
+		bs = make([]byte, 0, r.ContentLength)
+	}
+	buf := bytes.NewBuffer(bs)
+
+	_, err := buf.ReadFrom(r.Body)
+	if err != nil {
+		h.httpError(w, fmt.Sprintf("buckets - cannot read request body: %s", err.Error()), http.StatusBadRequest)
+		return
+	}
+
+	brd := &BucketsBody{}
+	if err = json.Unmarshal(buf.Bytes(), brd); err != nil {
+		h.httpError(w, fmt.Sprintf("buckets - cannot parse request body: %s", err.Error()), http.StatusBadRequest)
+		return
+	}
+	db, rp, err := bucket2dbrp(brd.Name)
```

Comment:
I think `rp == ""` should just be an error here - it would simplify the logic, be consistent with only having one transformation from name -> db/rp, and we can revisit later if we want it not to be an error.

We could still accept `brd.Rp` as either empty, or equal to the `rp` from the name.